### PR TITLE
MINOR Updated transaction index as optional in LogSegmentData.

### PR DESCRIPTION
**Pull request by satishd**
- Updated transaction index as optional in `LogSegmentData`.
 - Added a unit test for the introduced change.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by satishd** [comment]
@junrao : Pl take a look at this minor PR making transaction index as optional in `LogSegmentData` as we discussed earlier.

File: storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/LogSegmentData.java [hunk 1]
```diff
@@ -33,31 +34,32 @@
...
-     * @param txnIndex              transaction index file
+     * @param transactionIndex      transaction index file, which can be null
      * @param producerSnapshotIndex producer snapshot until this segment
      * @param leaderEpochIndex      leader-epoch-index until this segment
      */
     public LogSegmentData(Path logSegment,
                           Path offsetIndex,
                           Path timeIndex,
-                          Path txnIndex,
+                          Path transactionIndex,
```
Comment (junrao) [inline storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/LogSegmentData.java]: Could we make transactionIndex Optional<Path> to make it clear?

**Review by junrao — COMMENTED** [review]
@satishd : Thanks for the PR. Just a minor comment below.

File: storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/LogSegmentData.java (hunk 1 above)
Comment (satishd) [reply storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/LogSegmentData.java]: Passing `Optional` as arguments is not considered as a good practice. We still need to do null check for that Optional instance. 
[SO answer from Brian Goetz](https://stackoverflow.com/questions/26327957/should-java-8-getters-return-optional-type/26328555#26328555) mentioned the right usage of `Optional`. I have also updated PR not to use it as a field.

```
You should almost never use it as a field of something or a method parameter.
```

[Javadoc of Optional ](https://docs.oracle.com/javase/10/docs/api/java/util/Optional.html)suggests returning as an argument mentioned below.
```
API Note:
Optional is primarily intended for use as a method return type where there is a clear need to represent "no result," and where using null is likely to cause errors. A variable whose type is Optional should never itself be null; it should always point to an Optional instance.
```

Having said that, I do not have strong opinions on the above. I am fine with the conventions that we are following in this project if we have any on `Optional` usage.

Comment (ijuma) [reply storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/LogSegmentData.java]: The Java reasoning is pretty arbitrary and is inconsistent with the Scala recommendation (which we've been following and has worked pretty well for us). I suggest we stick with our approach.

File: storage/api/src/test/java/org/apache/kafka/server/log/remote/storage/LogSegmentDataTest.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        File dir = TestUtils.tempDirectory();
+        LogSegmentData logSegmentDataWithTransactionIndex = new LogSegmentData(
+                new File(dir, "log-segment").toPath(),
+                new File(dir, "offset-index").toPath(),
+                new File(dir, "time-index").toPath(),
+                new File(dir, "transaction-index").toPath(),
+                new File(dir, "producer-snapshot").toPath(),
+                ByteBuffer.allocate(1)
+        );
+        Assertions.assertTrue(logSegmentDataWithTransactionIndex.transactionIndex().isPresent());
```
Comment (ijuma) [inline storage/api/src/test/java/org/apache/kafka/server/log/remote/storage/LogSegmentDataTest.java:42]: Please use static imports to stick with the usual more concise assert style.

**Comment by satishd** [comment]
Thanks @junrao @ijuma for the review. Addressed the review comments with the latest commit.

**Review by junrao — COMMENTED** [review]
@satishd : Thanks for the updated PR. LGTM.

The scala 2.13 tests seem to have exited. Do you know why?

**Comment by satishd** [comment]
@junrao Thanks for the review.  
Tests failing/exiting is not only with scala-2.13. but the latest run failed with scala-2.12 also. `:core:integrationTest` is getting failed/exited in other PR jenkins jobs too. This failure does not seem to be related to this change.

**Comment by satishd** [comment]
@junrao I triggered a new build, there were a few intermittent test failures in scala 2.12/2.13 not related to the PR.

**Review by junrao — APPROVED** [review]
@satishd : Thanks for the reply. LGTM

**Pull request merged**